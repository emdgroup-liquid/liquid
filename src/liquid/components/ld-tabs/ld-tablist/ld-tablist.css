:host {
  /* colors */
  --ld-tablist-border-col: var(--ld-col-neutral-100);
  --ld-tablist-border-height: 0.1875rem;
  --ld-tablist-btn-scroll-col: var(--ld-col-neutral-900);
  --ld-tablist-btn-scroll-col-active: var(--ld-col-neutral-900);
  --ld-tablist-btn-scroll-bg-col: var(--ld-col-neutral-010);
  --ld-tablist-btn-scroll-bg-col-hover: var(--ld-col-neutral-100);
  --ld-tablist-btn-scroll-bg-col-active: var(--ld-col-neutral-300);

  /* theme colors */
  --ld-tablist-border-brand-col: var(--ld-thm-primary-hover);
  --ld-tablist-btn-scroll-bg-brand-col-active: var(--ld-thm-primary-active);
  --ld-tablist-btn-scroll-bg-brand-col-hover: var(--ld-thm-primary-hover);
  --ld-tablist-btn-scroll-bg-brand-col: var(--ld-thm-primary);

  /* theme colors (ld-tab) */
  --ld-tab-bg-brand-col-active: var(--ld-thm-primary-active);
  --ld-tab-bg-brand-col-focus: var(--ld-thm-primary-focus);
  --ld-tab-bg-brand-col-hover: var(--ld-thm-primary-hover);
  --ld-tab-bg-brand-col: var(--ld-thm-primary);
  --ld-tab-text-brand-col-focus: var(--ld-thm-primary-active);
  --ld-tab-disabled-bg-brand-col-focus: var(--ld-thm-primary-active);
  --ld-tab-disabled-text-brand-col-focus: var(--ld-thm-primary);
  --ld-tab-disabled-text-brand-col: var(--ld-thm-primary-focus);
  --ld-tab-selected-border-brand-col: var(--ld-thm-secondary);
}

.ld-tablist {
  position: relative;
  display: flex;
  overflow: hidden;
  z-index: 0;
  background-color: var(--ld-tablist-btn-scroll-bg-col);

  &::after {
    content: '';
    position: absolute;
    inset: auto 0 0 0;
    height: var(--ld-tablist-border-height);
    background-color: var(--ld-tablist-border-col);
    z-index: 2;
  }

  &--rounded-all {
    border-radius: var(--ld-br-m);
  }

  &--rounded-all-lg {
    border-radius: var(--ld-br-l);
  }

  &--rounded-top {
    border-radius: var(--ld-br-m) var(--ld-br-m) 0 0;
  }

  &--rounded-top-lg {
    border-radius: var(--ld-br-l) var(--ld-br-l) 0 0;
  }

  & ::slotted(ld-tab) {
    --ld-tab-indicator-height: var(--ld-tablist-border-height);
  }

  &--sm ::slotted(ld-tab) {
    --ld-tab-font: var(--ld-typo-body-s);
    --ld-tab-gap: 0.375rem;
    --ld-tab-min-height: 2rem;
    --ld-tab-padding-x: 0.625rem;
    --ld-tab-padding-y: 0.5rem;
  }

  &--lg ::slotted(ld-tab) {
    --ld-tab-font: var(--ld-typo-body-l);
    --ld-tab-gap: 0.875rem;
    --ld-tab-min-height: 3.125rem;
    --ld-tab-padding-x: 1.3125rem;
    --ld-tab-padding-y: 0.75rem;
  }

  &--ghost {
    --ld-tablist-btn-scroll-bg-col: transparent;
    ::slotted(ld-tab) {
      --ld-tab-bg-col: transparent;
    }
  }

  &--brand-color {
    /* ld-tablist internal vars */
    --ld-tablist-btn-scroll-col-active: var(--ld-col-wht);
    --ld-tablist-btn-scroll-col: var(--ld-col-wht);
    --ld-tablist-btn-scroll-bg-col: var(--ld-tablist-btn-scroll-bg-brand-col);
    --ld-tablist-btn-scroll-bg-col-hover: var(
      --ld-tablist-btn-scroll-bg-brand-col-hover
    );
    --ld-tablist-btn-scroll-bg-col-active: var(
      --ld-tablist-btn-scroll-bg-brand-col-active
    );
    --ld-tablist-border-col: var(--ld-tablist-border-brand-col);

    ::slotted(ld-tab) {
      --ld-tab-disabled-opacity: 0.5;
      --ld-tab-disabled-content-opacity: 1;

      /* colors */
      --ld-tab-selected-text-col-focus: var(--ld-col-wht);
      --ld-tab-selected-text-col: var(--ld-col-wht);
      --ld-tab-text-col: var(--ld-col-wht);

      /* theme colors */
      --ld-tab-bg-col: var(--ld-tab-bg-brand-col);
      --ld-tab-bg-col-focus: var(--ld-tab-bg-brand-col-focus);
      --ld-tab-bg-col-hover: var(--ld-tab-bg-brand-col-hover);
      --ld-tab-bg-col-active: var(--ld-tab-bg-brand-col-active);
      --ld-tab-disabled-bg-col-focus: var(--ld-tab-disabled-bg-brand-col-focus);
      --ld-tab-disabled-text-col-focus: var(
        --ld-tab-disabled-text-brand-col-focus
      );
      --ld-tab-disabled-text-col: var(--ld-tab-disabled-text-brand-col);
      --ld-tab-selected-border-col-focus: var(--ld-tab-selected-border-col);
      --ld-tab-selected-border-col: var(--ld-tab-selected-border-brand-col);
      --ld-tab-text-col-focus: var(--ld-tab-text-brand-col-focus);
    }
  }
}

.ld-tablist__scroll-container {
  display: flex;
  overflow-x: auto;
  scroll-snap-type: x mandatory;
  padding-bottom: 4rem;
  margin-bottom: -4rem;
}

.ld-tablist__btn-scroll {
  position: relative;
  border: 0;
  text-decoration: none;
  user-select: none;
  touch-action: manipulation;
  justify-content: center;
  align-items: center;
  -webkit-touch-callout: none;
  color: var(--ld-tablist-btn-scroll-col);
  background-color: var(--ld-tablist-btn-scroll-bg-col);
  box-shadow: var(--ld-shadow-active);
  width: var(--ld-sp-32);
  z-index: 2;

  &:focus:focus-visible {
    outline: none;
  }

  &[aria-disabled='true'] {
    svg {
      opacity: 0.25;
    }
  }

  &:where(:not([aria-disabled='true'])) {
    cursor: pointer;

    @media (hover: hover) {
      &:hover {
        background-color: var(--ld-tablist-btn-scroll-bg-col-hover);
      }
    }
    &:active:focus,
    &:active:focus:focus-visible {
      color: var(--ld-tablist-btn-scroll-col-active);
      background-color: var(--ld-tablist-btn-scroll-bg-col-active);
    }
  }

  &--left {
    left: 0;
  }
  &--right {
    right: 0;
  }
}
