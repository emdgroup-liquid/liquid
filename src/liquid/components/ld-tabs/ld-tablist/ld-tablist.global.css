:where(.ld-theme-ocean),
:where([class*='ld-theme'] .ld-theme-ocean) {
  --ld-tablist-border-brand-color: var(--ld-thm-ocean-bg-secondary);
  --ld-tablist-btn-scroll-bg-brand-color-active: var(--ld-col-rb-active);
  --ld-tablist-btn-scroll-bg-brand-color-hover: var(--ld-col-rb-hover);
  --ld-tablist-btn-scroll-bg-brand-color: var(--ld-thm-ocean-bg-primary);
  --ld-tab-bg-brand-color-active: var(--ld-col-rb-active);
  --ld-tab-bg-brand-color-focus-disabled: var(--ld-col-rb6);
  --ld-tab-bg-brand-color-focus: var(--ld-col-rb-focus);
  --ld-tab-bg-brand-color-hover: var(--ld-col-rb-hover);
  --ld-tab-bg-brand-color: var(--ld-thm-ocean-bg-primary);
  --ld-tab-border-brand-color-selected: var(--ld-thm-ocean-accent);
  --ld-tab-brand-color-disabled: var(--ld-col-rb3);
  --ld-tab-brand-color-focus-disabled: var(--ld-col-rb5);
  --ld-tab-brand-color-focus: var(--ld-col-rb6);
}

:where(.ld-theme-tea),
:where([class*='ld-theme'] .ld-theme-tea) {
  --ld-tablist-border-brand-color: var(--ld-thm-tea-bg-secondary);
  --ld-tablist-btn-scroll-bg-brand-color-active: var(--ld-col-rg-active);
  --ld-tablist-btn-scroll-bg-brand-color-hover: var(--ld-col-rg-hover);
  --ld-tablist-btn-scroll-bg-brand-color: var(--ld-thm-tea-bg-primary);
  --ld-tab-bg-brand-color-active: var(--ld-col-rg-active);
  --ld-tab-bg-brand-color-focus-disabled: var(--ld-col-rg6);
  --ld-tab-bg-brand-color-focus: var(--ld-col-rg-focus);
  --ld-tab-bg-brand-color-hover: var(--ld-col-rg-hover);
  --ld-tab-bg-brand-color: var(--ld-thm-tea-bg-primary);
  --ld-tab-border-brand-color-selected: var(--ld-thm-tea-accent);
  --ld-tab-brand-color-disabled: var(--ld-col-rg3);
  --ld-tab-brand-color-focus-disabled: var(--ld-col-rg5);
  --ld-tab-brand-color-focus: var(--ld-col-rg6);
}

:where(.ld-theme-bubblegum),
:where([class*='ld-theme'] .ld-theme-bubblegum) {
  --ld-tablist-border-brand-color: var(--ld-thm-bubblegum-bg-secondary);
  --ld-tablist-btn-scroll-bg-brand-color: var(--ld-thm-bubblegum-bg-primary);
  --ld-tablist-btn-scroll-bg-brand-color-hover: var(--ld-col-rp-hover);
  --ld-tablist-btn-scroll-bg-brand-color-active: var(--ld-col-rp-active);
  --ld-tab-bg-brand-color-active: var(--ld-col-rp-active);
  --ld-tab-bg-brand-color-focus-disabled: var(--ld-col-rp6);
  --ld-tab-bg-brand-color-focus: var(--ld-col-rp-focus);
  --ld-tab-bg-brand-color-hover: var(--ld-col-rp-hover);
  --ld-tab-bg-brand-color: var(--ld-thm-bubblegum-bg-primary);
  --ld-tab-border-brand-color-selected: var(--ld-thm-bubblegum-accent);
  --ld-tab-brand-color-disabled: var(--ld-col-rp3);
  --ld-tab-brand-color-focus-disabled: var(--ld-col-rp5);
  --ld-tab-brand-color-focus: var(--ld-col-rp6);
}

:where(.ld-theme-shake),
:where([class*='ld-theme'] .ld-theme-shake) {
  --ld-tablist-border-brand-color: var(--ld-thm-shake-bg-secondary);
  --ld-tablist-btn-scroll-bg-brand-color: var(--ld-thm-shake-bg-primary);
  --ld-tablist-btn-scroll-bg-brand-color-hover: var(--ld-col-rp-hover);
  --ld-tablist-btn-scroll-bg-brand-color-active: var(--ld-col-rp-active);
  --ld-tab-bg-brand-color-active: var(--ld-col-rp-active);
  --ld-tab-bg-brand-color-focus-disabled: var(--ld-col-rp6);
  --ld-tab-bg-brand-color-focus: var(--ld-col-rp-focus);
  --ld-tab-bg-brand-color-hover: var(--ld-col-rp-hover);
  --ld-tab-bg-brand-color: var(--ld-thm-shake-bg-primary);
  --ld-tab-border-brand-color-selected: var(--ld-thm-shake-accent);
  --ld-tab-brand-color-disabled: var(--ld-col-rp3);
  --ld-tab-brand-color-focus-disabled: var(--ld-col-rp5);
  --ld-tab-brand-color-focus: var(--ld-col-rp6);
}

:where(.ld-theme-solvent),
:where([class*='ld-theme'] .ld-theme-solvent) {
  --ld-tablist-border-brand-color: var(--ld-thm-solvent-bg-secondary);
  --ld-tablist-btn-scroll-bg-brand-color: var(--ld-thm-solvent-bg-primary);
  --ld-tablist-btn-scroll-bg-brand-color-hover: var(--ld-col-rp-hover);
  --ld-tablist-btn-scroll-bg-brand-color-active: var(--ld-col-rp-active);
  --ld-tab-bg-brand-color-active: var(--ld-col-rp-active);
  --ld-tab-bg-brand-color-focus-disabled: var(--ld-col-rp6);
  --ld-tab-bg-brand-color-focus: var(--ld-col-rp-focus);
  --ld-tab-bg-brand-color-hover: var(--ld-col-rp-hover);
  --ld-tab-bg-brand-color: var(--ld-thm-solvent-bg-primary);
  --ld-tab-border-brand-color-selected: var(--ld-thm-solvent-accent);
  --ld-tab-brand-color-disabled: var(--ld-col-rp3);
  --ld-tab-brand-color-focus-disabled: var(--ld-col-rp5);
  --ld-tab-brand-color-focus: var(--ld-col-rp6);
}
