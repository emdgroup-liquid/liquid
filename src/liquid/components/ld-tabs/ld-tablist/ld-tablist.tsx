import '../../../components' // type definitions for type checks and intelliSense
import { Component, Element, h, Host, Listen, Prop, State } from '@stencil/core'
import { LdTab } from '../ld-tab/ld-tab'
import { getClassNames } from 'src/liquid/utils/getClassNames'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part arrow - Both arrow button elements
 * @part arrow-icon - Both arrow icon SVGs
 * @part arrow-icon-left - Left arrow icon SVG
 * @part arrow-icon-right - Right arrow icon SVG
 * @part arrow-left - Left arrow button element
 * @part arrow-right - Right arrow button element
 * @part scroll-container - Container wrapping the slot
 * @part wrapper - Container wrapping the arrows and the scroll-container
 */
@Component({
  tag: 'ld-tablist',
  styleUrl: 'ld-tablist.css',
  shadow: true,
})
export class LdTablist {
  @Element() el: HTMLElement

  /** Size of the tabs. */
  @Prop() size?: 'sm' | 'lg'

  /** Display mode. */
  @Prop() mode?: 'ghost' | 'brand-color'

  /** Sets border radii. */
  @Prop() rounded?: 'all' | 'all-lg' | 'top' | 'top-lg'

  private slotContainerRef!: HTMLElement
  private btnScrollLeftRef!: HTMLButtonElement

  @State() scrollable: boolean
  @State() scrollLeftEnabled: boolean
  @State() scrollRightEnabled: boolean

  @Listen('resize', { target: 'window', passive: true })
  handleWindowResize() {
    this.updateScrollable()
    this.updateScrollButtons()
  }

  private updateScrollable() {
    const scrollButtonsWidth =
      2 * this.btnScrollLeftRef.getBoundingClientRect().width
    const scrollContainerWidth = this.slotContainerRef.getBoundingClientRect()
      .width
    const contentWidth = Array.from(this.el.children)
      .map((child) => child.getBoundingClientRect().width)
      .reduce((a, b) => a + b)
    this.scrollable =
      scrollContainerWidth + (this.scrollable ? scrollButtonsWidth : 0) <
      contentWidth
  }

  private updateScrollButtons() {
    if (!this.scrollable) return
    this.scrollLeftEnabled = this.slotContainerRef.scrollLeft > 0
    this.scrollRightEnabled =
      this.slotContainerRef.scrollLeft +
        this.slotContainerRef.getBoundingClientRect().width -
        this.slotContainerRef.scrollWidth <
      0
  }

  private scroll(dir: 'left' | 'right') {
    this.slotContainerRef.scrollTo({
      left:
        this.slotContainerRef.scrollLeft +
        ((dir === 'left' ? -1 : 1) *
          this.slotContainerRef.getBoundingClientRect().width) /
          2,
      behavior: 'smooth',
    })
  }

  private focusTab(prevLdTab: HTMLElement, dir: 'left' | 'right') {
    const currentTab =
      dir === 'left'
        ? prevLdTab.previousElementSibling
        : prevLdTab.nextElementSibling
    if (currentTab) {
      ;((currentTab as unknown) as LdTab).focusInner()
      currentTab.scrollIntoView({
        behavior: 'smooth',
        block: 'nearest',
        inline: 'center',
      })
    }
  }

  private setFocusOnSelectedTabpanel() {
    // TODO: fix Stencils DOM implementation for unit testing and replace
    // ;(this.el
    //   .closest('ld-tabs')
    //   .querySelector('ld-tabpanel:not([hidden])') as HTMLElement)?.focus()
    Array.from(this.el.closest('ld-tabs').querySelectorAll('ld-tabpanel'))
      .find((tabpanel) => !tabpanel.hasAttribute('hidden'))
      ?.focus()
  }

  private onKeydown(ev) {
    switch (ev.key) {
      case 'ArrowLeft':
        ev.preventDefault()
        this.focusTab(ev.target, 'left')
        return
      case 'ArrowRight': {
        ev.preventDefault()
        this.focusTab(ev.target, 'right')
        return
      }
      case 'ArrowDown': {
        ev.preventDefault()
        this.setFocusOnSelectedTabpanel()
        return
      }
    }
  }

  componentWillLoad() {
    this.el.querySelectorAll('ld-icon').forEach((icon) => {
      if (this.size !== undefined) {
        icon.setAttribute('size', this.size)
      } else {
        icon.removeAttribute('size')
      }
    })
    this.el.querySelectorAll('.ld-icon').forEach((icon) => {
      icon.classList.remove('ld-icon--sm', 'ld-icon--lg')
      if (this.size === 'sm') {
        icon.classList.remove('ld-icon--lg')
        icon.classList.add('ld-icon--sm')
      } else if (this.size === 'lg') {
        icon.classList.remove('ld-icon--sm')
        icon.classList.add('ld-icon--lg')
      } else {
        icon.classList.remove('ld-icon--sm', 'ld-icon--lg')
      }
    })
  }

  componentDidLoad() {
    setTimeout(() => {
      this.updateScrollable()
      this.updateScrollButtons()
    })
  }

  render() {
    return (
      <Host onKeydown={this.onKeydown.bind(this)} role="tablist">
        <div
          class={getClassNames([
            'ld-tablist',
            this.size && `ld-tablist--${this.size}`,
            this.mode && `ld-tablist--${this.mode}`,
            this.rounded && `ld-tablist--rounded-${this.rounded}`,
          ])}
          part="wrapper"
        >
          <button
            aria-disabled={this.scrollLeftEnabled ? undefined : 'true'}
            class="ld-tablist__btn-scroll ld-tablist__btn-scroll--left"
            hidden={!this.scrollable}
            onClick={this.scroll.bind(this, 'left')}
            part="arrow arrow-left"
            ref={(el) => (this.btnScrollLeftRef = el)}
            tabindex="-1"
          >
            <svg
              fill="none"
              height="16"
              part="arrow-icon arrow-icon-left"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>Scroll left</title>
              <path
                d="M10 13L6 8L10 3"
                stroke="currentColor"
                stroke-width="2"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <div
            class="ld-tablist__scroll-container"
            onScroll={this.updateScrollButtons.bind(this)}
            part="scroll-container"
            ref={(el) => (this.slotContainerRef = el)}
          >
            <slot></slot>
          </div>
          <button
            aria-disabled={this.scrollRightEnabled ? undefined : 'true'}
            class="ld-tablist__btn-scroll ld-tablist__btn-scroll--right"
            hidden={!this.scrollable}
            onClick={this.scroll.bind(this, 'right')}
            part="arrow arrow-right"
            tabindex="-1"
          >
            <svg
              fill="none"
              height="16"
              part="arrow-icon arrow-icon-right"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>Scroll right</title>
              <path
                d="M6 13L10 8L6 3"
                stroke="currentColor"
                stroke-width="2"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
        </div>
      </Host>
    )
  }
}
