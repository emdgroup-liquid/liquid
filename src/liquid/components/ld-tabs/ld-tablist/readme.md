---
eleventyNavigation:
  key: Tab List
  parent: Tabs
layout: layout.njk
title: Tab List
permalink: components/ld-tabs/ld-tablist/
---

# ld-tablist

The `ld-tablist` component is a subcomponent for `ld-tabs`.

Please refer to the [`ld-tabs` documentation](components/ld-tabs) for usage examples.

---


<!-- Auto Generated Below -->


## Properties

| Property  | Attribute | Description                                              | Type                                     | Default     |
| --------- | --------- | -------------------------------------------------------- | ---------------------------------------- | ----------- |
| `key`     | `key`     | for tracking the node's identity when working with lists | `string \| number`                       | `undefined` |
| `mode`    | `mode`    | Display mode.                                            | `"brand-color" \| "ghost"`               | `undefined` |
| `ref`     | `ref`     | reference to component                                   | `any`                                    | `undefined` |
| `rounded` | `rounded` | Sets border radii.                                       | `"all" \| "all-lg" \| "top" \| "top-lg"` | `undefined` |
| `size`    | `size`    | Size of the tabs.                                        | `"lg" \| "sm"`                           | `undefined` |


## Shadow Parts

| Part                 | Description                                            |
| -------------------- | ------------------------------------------------------ |
| `"arrow"`            | Both arrow button elements                             |
| `"arrow-icon"`       | Both arrow icon SVGs                                   |
| `"arrow-icon-left"`  | Left arrow icon SVG                                    |
| `"arrow-icon-right"` | Right arrow icon SVG                                   |
| `"arrow-left"`       | Left arrow button element                              |
| `"arrow-right"`      | Right arrow button element                             |
| `"scroll-container"` | Container wrapping the slot                            |
| `"wrapper"`          | Container wrapping the arrows and the scroll-container |


----------------------------------------------

 
