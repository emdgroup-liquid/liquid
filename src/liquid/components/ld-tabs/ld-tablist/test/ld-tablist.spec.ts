import { newSpecPage } from '@stencil/core/testing'
import { LdTablist } from '../ld-tablist'
import { LdTab } from '../../ld-tab/ld-tab'
import '../../../../utils/resizeObserver'
import { getTriggerableMutationObserver } from '../../../../utils/mutationObserver'
import { LdIcon } from '../../../ld-icon/ld-icon'

const components = [LdTablist, LdTab, LdIcon]

describe('ld-tablist', () => {
  beforeEach(() => {
    jest.clearAllTimers()
  })

  describe('modifiers', () => {
    it('size', async () => {
      const page = await newSpecPage({
        components,
        html: `
        <ld-tablist size="sm">
          <ld-tab selected>Fruits</ld-tab>
          <ld-tab><ld-icon name="placeholder" />Vegetables</ld-tab>
        </ld-tablist>
      `,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('watches size', async () => {
      const page = await newSpecPage({
        components,
        html: `
        <ld-tablist size="sm">
          <ld-tab selected>Fruits</ld-tab>
          <ld-tab><ld-icon name="placeholder" />Vegetables</ld-tab>
          <ld-tab><span class="ld-icon" />Nuts</ld-tab>
        </ld-tablist>
      `,
      })
      await page.waitForChanges()

      const ldIcon = page.root.querySelector('ld-icon')
      const icon = page.root.querySelector('.ld-icon')
      expect(ldIcon.size).toEqual('sm')
      expect(icon).toHaveClass('ld-icon--sm')

      page.root.size = 'lg'
      page.waitForChanges()
      expect(ldIcon.size).toEqual('lg')
      expect(icon).not.toHaveClass('ld-icon--sm')
      expect(icon).toHaveClass('ld-icon--lg')

      page.root.size = undefined
      page.waitForChanges()
      expect(ldIcon.size).toEqual(undefined)
      expect(icon).not.toHaveClass('ld-icon--sm')
      expect(icon).not.toHaveClass('ld-icon--lg')
    })

    it('mode', async () => {
      const page = await newSpecPage({
        components,
        html: `
        <ld-tablist mode="ghost">
          <ld-tab selected>Fruits</ld-tab>
          <ld-tab>Vegetables</ld-tab>
        </ld-tablist>
      `,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('rounded', async () => {
      const page = await newSpecPage({
        components,
        html: `
        <ld-tablist rounded="all-lg">
          <ld-tab selected>Fruits</ld-tab>
          <ld-tab>Vegetables</ld-tab>
        </ld-tablist>
      `,
      })
      expect(page.root).toMatchSnapshot()
    })
  })

  describe('floating mode', () => {
    it('adds and removes focus visible class on click and focus out', async () => {
      const page = await newSpecPage({
        components,
        html: `
        <ld-tablist mode="floating">
          <ld-tab selected>Fruits</ld-tab>
          <ld-tab>Vegetables</ld-tab>
        </ld-tablist>
      `,
      })
      const tabList = page.root.shadowRoot.querySelector('.ld-tablist')
      expect(tabList).toHaveClass('ld-tablist--focus-visible')

      const evClick = {
        type: 'click',
        pointerType: 'mouse',
      } as unknown as MouseEvent
      page.root.dispatchEvent(evClick)
      await page.waitForChanges()
      expect(tabList).not.toHaveClass('ld-tablist--focus-visible')

      const evFocus = {
        type: 'focusout',
        relatedTarget: page.body,
      } as unknown as FocusEvent
      page.root.dispatchEvent(evFocus)
      await page.waitForChanges()
      expect(tabList).toHaveClass('ld-tablist--focus-visible')
    })

    it('updates selected tab indicator', async () => {
      const page = await newSpecPage({
        components,
        html: `
        <ld-tablist mode="floating">
          <ld-tab selected>Fruits</ld-tab>
          <ld-tab>Vegetables</ld-tab>
        </ld-tablist>
      `,
      })
      const tabList = page.root.shadowRoot.querySelector('.ld-tablist')
      expect(tabList).toHaveClass('ld-tablist--focus-visible')

      const indicator = page.root.shadowRoot.querySelector(
        '.ld-tablist__active-tab-indicator'
      )

      page.root
        .querySelectorAll('ld-tab')[1]
        .dispatchEvent(new CustomEvent('ldtabselect', { bubbles: true }))
      await page.waitForChanges()

      expect((indicator as HTMLElement).style.opacity).toEqual('1')

      page.root.querySelectorAll('ld-tab')[0].remove()
      getTriggerableMutationObserver().trigger([])
      page.waitForChanges()
      expect((indicator as HTMLElement).style.opacity).toEqual('0')
    })
  })

  it('sets initialized class for enabling transitions after initialization', async () => {
    const page = await newSpecPage({
      components,
      html: `
        <ld-tablist mode="floating">
          <ld-tab selected>Fruits</ld-tab>
          <ld-tab>Vegetables</ld-tab>
        </ld-tablist>
      `,
    })
    const tabList = page.root.shadowRoot.querySelector('.ld-tablist')
    expect(tabList).not.toHaveClass('ld-tablist--initialized')
    jest.advanceTimersByTime(0)
    await page.waitForChanges()
    expect(tabList).toHaveClass('ld-tablist--initialized')
  })
})
