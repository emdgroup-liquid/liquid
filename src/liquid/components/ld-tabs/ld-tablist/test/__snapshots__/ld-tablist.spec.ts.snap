// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-tablist modifiers mode 1`] = `
<ld-tablist mode="ghost" role="tablist">
  <mock:shadow-root>
    <div class="ld-tablist ld-tablist--ghost" part="wrapper">
      <button aria-disabled="true" class="ld-tablist__btn-scroll ld-tablist__btn-scroll--left" hidden="" part="arrow arrow-left" tabindex="-1">
        <svg fill="none" height="16" part="arrow-icon arrow-icon-left" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg">
          <title>
            Scroll left
          </title>
          <path d="M10 13L6 8L10 3" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"></path>
        </svg>
      </button>
      <div class="ld-tablist__scroll-container" part="scroll-container">
        <slot></slot>
      </div>
      <button aria-disabled="true" class="ld-tablist__btn-scroll ld-tablist__btn-scroll--right" hidden="" part="arrow arrow-right" tabindex="-1">
        <svg fill="none" height="16" part="arrow-icon arrow-icon-right" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg">
          <title>
            Scroll right
          </title>
          <path d="M6 13L10 8L6 3" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"></path>
        </svg>
      </button>
    </div>
  </mock:shadow-root>
  <ld-tab selected="">
    <mock:shadow-root>
      <button aria-selected="true" class="ld-tab" part="button focusable" role="tab">
        <span class="ld-tab__spacer" part="spacer spacer-left"></span>
        <span class="ld-tab__content" part="content">
          <slot></slot>
        </span>
        <span class="ld-tab__spacer" part="spacer spacer-right"></span>
      </button>
    </mock:shadow-root>
    Fruits
  </ld-tab>
  <ld-tab>
    <mock:shadow-root>
      <button class="ld-tab" part="button focusable" role="tab" tabindex="-1">
        <span class="ld-tab__spacer" part="spacer spacer-left"></span>
        <span class="ld-tab__content" part="content">
          <slot></slot>
        </span>
        <span class="ld-tab__spacer" part="spacer spacer-right"></span>
      </button>
    </mock:shadow-root>
    Vegetables
  </ld-tab>
</ld-tablist>
`;

exports[`ld-tablist modifiers rounded 1`] = `
<ld-tablist role="tablist" rounded="all-lg">
  <mock:shadow-root>
    <div class="ld-tablist ld-tablist--rounded-all-lg" part="wrapper">
      <button aria-disabled="true" class="ld-tablist__btn-scroll ld-tablist__btn-scroll--left" hidden="" part="arrow arrow-left" tabindex="-1">
        <svg fill="none" height="16" part="arrow-icon arrow-icon-left" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg">
          <title>
            Scroll left
          </title>
          <path d="M10 13L6 8L10 3" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"></path>
        </svg>
      </button>
      <div class="ld-tablist__scroll-container" part="scroll-container">
        <slot></slot>
      </div>
      <button aria-disabled="true" class="ld-tablist__btn-scroll ld-tablist__btn-scroll--right" hidden="" part="arrow arrow-right" tabindex="-1">
        <svg fill="none" height="16" part="arrow-icon arrow-icon-right" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg">
          <title>
            Scroll right
          </title>
          <path d="M6 13L10 8L6 3" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"></path>
        </svg>
      </button>
    </div>
  </mock:shadow-root>
  <ld-tab selected="">
    <mock:shadow-root>
      <button aria-selected="true" class="ld-tab" part="button focusable" role="tab">
        <span class="ld-tab__spacer" part="spacer spacer-left"></span>
        <span class="ld-tab__content" part="content">
          <slot></slot>
        </span>
        <span class="ld-tab__spacer" part="spacer spacer-right"></span>
      </button>
    </mock:shadow-root>
    Fruits
  </ld-tab>
  <ld-tab>
    <mock:shadow-root>
      <button class="ld-tab" part="button focusable" role="tab" tabindex="-1">
        <span class="ld-tab__spacer" part="spacer spacer-left"></span>
        <span class="ld-tab__content" part="content">
          <slot></slot>
        </span>
        <span class="ld-tab__spacer" part="spacer spacer-right"></span>
      </button>
    </mock:shadow-root>
    Vegetables
  </ld-tab>
</ld-tablist>
`;

exports[`ld-tablist modifiers size 1`] = `
<ld-tablist role="tablist" size="sm">
  <mock:shadow-root>
    <div class="ld-tablist ld-tablist--sm" part="wrapper">
      <button aria-disabled="true" class="ld-tablist__btn-scroll ld-tablist__btn-scroll--left" hidden="" part="arrow arrow-left" tabindex="-1">
        <svg fill="none" height="16" part="arrow-icon arrow-icon-left" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg">
          <title>
            Scroll left
          </title>
          <path d="M10 13L6 8L10 3" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"></path>
        </svg>
      </button>
      <div class="ld-tablist__scroll-container" part="scroll-container">
        <slot></slot>
      </div>
      <button aria-disabled="true" class="ld-tablist__btn-scroll ld-tablist__btn-scroll--right" hidden="" part="arrow arrow-right" tabindex="-1">
        <svg fill="none" height="16" part="arrow-icon arrow-icon-right" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg">
          <title>
            Scroll right
          </title>
          <path d="M6 13L10 8L6 3" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"></path>
        </svg>
      </button>
    </div>
  </mock:shadow-root>
  <ld-tab selected="">
    <mock:shadow-root>
      <button aria-selected="true" class="ld-tab" part="button focusable" role="tab">
        <span class="ld-tab__spacer" part="spacer spacer-left"></span>
        <span class="ld-tab__content" part="content">
          <slot></slot>
        </span>
        <span class="ld-tab__spacer" part="spacer spacer-right"></span>
      </button>
    </mock:shadow-root>
    Fruits
  </ld-tab>
  <ld-tab>
    <mock:shadow-root>
      <button class="ld-tab" part="button focusable" role="tab" tabindex="-1">
        <span class="ld-tab__spacer" part="spacer spacer-left"></span>
        <span class="ld-tab__content" part="content">
          <slot></slot>
        </span>
        <span class="ld-tab__spacer" part="spacer spacer-right"></span>
      </button>
    </mock:shadow-root>
    Vegetables
  </ld-tab>
</ld-tablist>
`;
